signature/cose/envelope.go

```diff
@@ -0,0 +1,587 @@
+	}
+}
+
+// ParseEnvelope parses envelopeBytes to a COSE signature envelope.
+func ParseEnvelope(envelopeBytes []byte) (signature.Envelope, error) {
+	var msg cose.Sign1Message
+	if err := msg.UnmarshalCBOR(envelopeBytes); err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	return &base.Envelope{
+		Envelope: &envelope{
+			base: &msg,
+		},
+		Raw: envelopeBytes,
+	}, nil
+}
+
+// Sign implements signature.Envelope interface.
+// On success, this function returns the COSE signature envelope byte slice.
+func (e *envelope) Sign(req *signature.SignRequest) ([]byte, error) {
+	// get built-in signer from go-cose or remote signer based on req.Signer
+	signer, err := getSigner(req.Signer)
+	if err != nil {
+		return nil, &signature.InvalidSignRequestError{Msg: err.Error()}
+	}
+
+	// prepare COSE_Sign1 message
+	msg := cose.NewSign1Message()
+
+	// generate protected headers of COSE envelope
+	msg.Headers.Protected.SetAlgorithm(signer.Algorithm())
+	if err := generateProtectedHeaders(req, msg.Headers.Protected); err != nil {
+		return nil, &signature.InvalidSignRequestError{Msg: err.Error()}
+	}
+
+	// generate payload of COSE envelope
+	msg.Headers.Protected[cose.HeaderLabelContentType] = req.Payload.ContentType
+	msg.Payload = req.Payload.Content
+
+	// core sign process, generate signature of COSE envelope
+	if err := msg.Sign(rand.Reader, nil, signer); err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+
+	// generate unprotected headers of COSE envelope
+	generateUnprotectedHeaders(req, signer, msg.Headers.Unprotected)
+
+	// TODO: needs to add headerKeyTimeStampSignature here.
+
+	// encode Sign1Message into COSE_Sign1_Tagged object
+	encoded, err := msg.MarshalCBOR()
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	e.base = msg
+	return encoded, nil
+}
+
+// Verify implements signature.Envelope interface.
+// Note: Verfiy only verifies integrity of the given COSE envelope
+func (e *envelope) Verify() (*signature.EnvelopeContent, error) {
+	// sanity check
+	if e.base == nil {
+		return nil, &signature.InvalidSignatureError{Msg: "missing COSE signature envelope"}
+	}
+
+	certs, ok := e.base.Headers.Unprotected[cose.HeaderLabelX5Chain].([]interface{})
+	if !ok || len(certs) == 0 {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope invalid certificate chain"}
+	}
+	certRaw, ok := certs[0].([]byte)
+	if !ok {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope malformed leaf certificate"}
+	}
+	cert, err := x509.ParseCertificate(certRaw)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+
+	// core verify process, verify integrity of COSE envelope
+	publicKeyAlg, err := getSignatureAlgorithm(cert)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	verifier, err := cose.NewVerifier(publicKeyAlg, cert.PublicKey)
```

Comment:
We should add validation around supported algorithms? [veraison/go-cose](https://github.com/veraison/go-cose/blob/main/verifier.go) supports `AlgorithmEd25519` which notary doesnt and if go-cose might add support for symmetric keys that would become a major security issue for notation.

Reply:
Yes, 	publicKeyAlg, err := getSignatureAlgorithm(cert) checks the algorithm, we only support RSA and EC.